### ISSUE TITLE
"error[internal]: Resolve error" When bundling packages with optional dependencies.

### ISSUE THREAD

**Issue by ConnorMinielly** [web-infra-dev/rspack#2986]
### System Info

  System:
    OS: macOS 13.2.1
    CPU: (10) arm64 Apple M1 Pro
    Memory: 510.44 MB / 32.00 GB
    Shell: 3.5.1 - /opt/homebrew/bin/fish
  Binaries:
    Node: 16.19.0 - /var/folders/d9/hypnz_jd6rz0987fb9xm69_00000gn/T/yarn--1682543747170-0.2739198567650005/node
    Yarn: 1.22.19 - /var/folders/d9/hypnz_jd6rz0987fb9xm69_00000gn/T/yarn--1682543747170-0.2739198567650005/yarn
    npm: 8.19.3 - /opt/homebrew/opt/nvm/versions/node/v16.19.0/bin/npm
  Browsers:
    Chrome: 112.0.5615.137
    Firefox Developer Edition: 110.0
    Safari: 16.3
  npmPackages:
    @rspack/cli: ^0.1.9 => 0.1.9 
    @rspack/plugin-minify: ^0.1.9 => 0.1.9 
    @rspack/plugin-node-polyfill: ^0.1.9 => 0.1.9 

## TLDR:
Bundling dependencies that have optional sub-dependencies causes errors!

## Background:
I'm attempting to update some somewhat-outdated build steps that currently use webpack to rspack. These builds are targeted to node, simply explained were bundling server functions for deployment to AWS lambda so as to minimize code size (we're also transpiling Typescript to JS). Currently though the builds are way slower than they need to be as they're just webpack with the standard ts-loader. I've tried moving to the swc-loader but something about that approach is missing the mark and causing inexplicable runtime errors. Having discovered rspack, it seems like a possible solution however I'm encountering a very strange issue: Dependencies that have optional dependencies are causing build errors. 

## The Problem:
When I execute `rspack build` I get a couple dozen errors like the following:
```
error[internal]: Resolve error
   ┌─ [PATH TO NODE_MODULES]/node_modules/@aws-sdk/signature-v4-multi-region/dist-es/SignatureV4MultiRegion.js:27:64
   │
27 │                 CrtSignerV4 = typeof require === "function" && require("@aws-sdk/signature-v4-crt").CrtSignerV4;
   │                                                                ^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^ Failed to resolve @aws-sdk/signature-v4-crt in [PATH TO NODE_MODULES]/node_modules/@aws-sdk/signature-v4-multi-region/dist-es/SignatureV4MultiRegion.js



error[internal]: Resolve error
  ┌─ [PATH TO NODE_MODULES]/node_modules/ws/lib/validation.js:4:23
  │
4 │   const isValidUTF8 = require('utf-8-validate');
  │                       ^^^^^^^^^^^^^^^^^^^^^^^^^ Failed to resolve utf-8-validate in [PATH TO NODE_MODULES]/node_modules/ws/lib/validation.js



error[internal]: Resolve error
    ┌─ [PATH TO NODE_MODULES]/node_modules/ws/lib/buffer-util.js:105:22
    │
105 │   const bufferUtil = require('bufferutil');
    │                      ^^^^^^^^^^^^^^^^^^^^^ Failed to resolve bufferutil in [PATH TO NODE_MODULES]/node_modules/ws/lib/buffer-util.js
```

As far as I can tell the through line is that all of these sub-dependencies are optional peer deps. For instance `bufferutil` and  and `utf-8-validate` are defined in the package.json form ws as:
```
"peerDependencies": {
    "bufferutil": "^4.0.1",
    "utf-8-validate": "^5.0.2"
  },
  "peerDependenciesMeta": {
    "bufferutil": {
      "optional": true
    },
    "utf-8-validate": {
      "optional": true
    }
  },
```
rspack should be gracefully handling the optional/dynamic dependencies and not outputting tons of errors. I can confirm that manually installing the dependencies does resolve the errors, but one shouldn't have to install every single optional dependency. 

Again, these builds are fully working with webpack + ts-loader and while I have other issues with build tools like esbuild (namely that it doesn't support decorators) this is definitely not one of them. 

Is there a configuration I'm missing to allow optional deps?

**Also, important note: ** The build command does in fact still output a bundled file despite the long list of errors, which is very interesting. I'm still working to figure out if the output is valid or not, or if it results in runtime errors.

## rspack config:
```typescript
import {Configuration} from '@rspack/cli';
import PluginPolyfill from '@rspack/plugin-node-polyfill';
import PluginMinify from '@rspack/plugin-minify';

import path from 'path';

const config: Configuration = {
  target: 'node',
  mode: 'development',
  entry: './src/server.ts',
  plugins: [new PluginPolyfill()],
  output: {
    path: path.resolve(__dirname, 'build'),
    filename: 'server.js',
    library: {
      type: 'commonjs2',
    },
  },
  optimization: {
    minimizer: [
      new PluginMinify({
        minifier: 'terser',
      }),
    ],
  },
};

// @ts-ignore
export = config;

```

### Reproduce link

https://stackblitz.com/edit/rspack-error-repro-cm

### Reproduce Steps

NOTE: rspack won't work in stackblitz due to native addons, but you can download the reproduction repo then:

1. run `npm i`
2. run `npm run build`
3. see error output (build folder still get created)

**Comment by hardfist** [comment]
>  rspack should be gracefully handling the optional/dynamic dependencies and not outputting tons of errors. I can confirm that manually installing the dependencies does resolve the errors, but one shouldn't have to install every single optional dependency.

it seems webpack doesn't do this automatically and we should align with webpack about this behavior, so you need to external optionalDependencies manually
